Stripe/STPTelemetryClient.m

```diff
@@ -0,0 +1,190 @@
+        [webView evaluateJavaScript:@"navigator.userAgent" completionHandler:^(id result, __unused NSError *error) {
+            self.userAgent = result;
+        }];
+        _webView = webView;
+    }
+    return self;
+}
+
+- (void)dealloc {
+    [[NSNotificationCenter defaultCenter] removeObserver:self];
+}
+
+- (void)addTelemetryFieldsToParams:(NSMutableDictionary *)params {
+    params[@"muid"] = [self muid];
+    params[@"time_on_page"] = [self timeOnPage];
+}
+
+- (void)applicationDidBecomeActive {
+    self.appOpenTime = [NSDate date];
+}
+
+- (NSString *)muid {
+    NSString *muid = [[[UIDevice currentDevice] identifierForVendor] UUIDString];
+    return muid ?: @"";
+}
+
+- (NSNumber *)timeOnPage {
+    if (!self.appOpenTime) {
+        return @(0);
+    }
+    NSTimeInterval seconds = [[NSDate date] timeIntervalSinceDate:self.appOpenTime];
+    NSInteger millis = (NSInteger)round(seconds*1000);
+    return @(MAX(millis, 0));
+}
+
+- (NSString *)language {
+    NSString *localeID = [[NSLocale currentLocale] localeIdentifier];
+    return localeID ?: @"";
+}
+
+- (NSString *)platform {
+    struct utsname systemInfo;
+    uname(&systemInfo);
+    NSString *deviceType = @(systemInfo.machine);
+    return deviceType ?: @"";
+}
+
+- (NSString *)screenSize {
+    UIScreen *screen = [UIScreen mainScreen];
+    CGRect screenRect = [screen bounds];
+    CGFloat width = screenRect.size.width;
+    CGFloat height = screenRect.size.height;
+    CGFloat scale = [screen scale];
+    return [NSString stringWithFormat:@"%.0fw_%.0fh_%.0fr", width, height, scale];
+}
+
+- (NSString *)timeZoneOffset {
+    NSTimeZone *timeZone = [NSTimeZone localTimeZone];
+    double hoursFromGMT = (double)timeZone.secondsFromGMT/(60*60);
+    return [NSString stringWithFormat:@"%.0f", hoursFromGMT];
+}
+
+- (NSString *)userAgent {
+    return _userAgent ?: @"";
+}
+
+- (NSDictionary *)payload {
+    NSMutableDictionary *payload = [NSMutableDictionary new];
+    NSMutableArray *fields = [NSMutableArray new];
+    for (STPMetricField i = 0; i < STPMetricFieldMax; i++) {
+        switch (i) {
+            case STPMetricFieldLanguage:
+                [fields addObject:@[[self language]]];
+                break;
+            case STPMetricFieldPlatform:
+                [fields addObject:@[[self platform]]];
+                break;
+            case STPMetricFieldScreenSize:
+                [fields addObject:@[[self screenSize]]];
+                break;
+            case STPMetricFieldTimeZoneOffset:
+                [fields addObject:@[[self timeZoneOffset]]];
+                break;
+            case STPMetricFieldUserAgent:
+                [fields addObject:@[[self userAgent]]];
+                break;
+            default:
+                [fields addObject:@[@""]];
+                break;
+        }
+    }
+    payload[@"f"] = fields;
+    payload[@"d"] = @[
+                      @"",
+                      @{@"fr": @"",
```

Comment:
Same here. The `fr` and `ft` params can be non-existent rather than being empty strings. The first param above does need to be an empty string though, just to keep the positions straight.